albatross/core/dataset.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_CORE_DATASET_H
+#define ALBATROSS_CORE_DATASET_H
+
+#include "core/distribution.h"
+#include "core/traits.h"
+#include <Eigen/Core>
+#include <cereal/archives/json.hpp>
+#include <map>
+#include <vector>
+
+namespace albatross {
+
+// A JointDistribution has a dense covariance matrix, which
+// contains the covariance between each variable and all others.
+using JointDistribution = Distribution<Eigen::MatrixXd>;
+
+// We use a wrapper around DiagonalMatrix in order to make
+// the resulting distribution serializable
+using DiagonalMatrixXd =
+    Eigen::SerializableDiagonalMatrix<double, Eigen::Dynamic>;
+// A MarginalDistribution has only a digaonal covariance
+// matrix, so in turn only describes the variance of each
+// variable independent of all others.
+using MarginalDistribution = Distribution<DiagonalMatrixXd>;
+
+/*
+ * A RegressionDataset holds two vectors of data, the features
+ * where a single feature can be any class that contains the information used
+ * to make predictions of the target.  This is called a RegressionDataset since
+ * it is assumed that each feature is regressed to a single double typed
+ * target.
+ */
+template <typename FeatureType> struct RegressionDataset {
+  std::vector<FeatureType> features;
+  MarginalDistribution targets;
+  std::map<std::string, std::string> metadata;
+
+  RegressionDataset(){};
+
+  RegressionDataset(const std::vector<FeatureType> &features_,
```

Comment:
the trailing underscore convention is backwards here from what we normally have. I'm going to assume its consistent with the rest of albatross however which is all that really matters. I think we talked about this ages ago... :)

Reply:
yeah, the rest of albatross uses a trailing underscore to avoid shadowing.  I started to try to change them all ... then decided it was more work than it was worth.